### PR TITLE
Fix building docs on Windows

### LINKED ISSUE #6625: Unable to use the docsrs configuration under Windows
**Version**
1.38.0

**Platform**
Windows 11 Pro / 64 bit

**Description**
On Windows, the documentation for a project using Tokio cannot be build with `docsrs` enabled.  This makes it difficult-to-impossible to include the automatically generated feature flag details.  More details are available from [this post on the Rust Users forum](https://users.rust-lang.org/t/tokio-rustflags-cfg-docsrs-trouble/112683).

A test project was created with `cargo new tokio-issue-112683`.

Cargo.toml was updated to this...
``` toml
[package]
name = "tokio-issue-112683"
version = "0.1.0"
edition = "2021"

[dependencies]
tokio = { version = "1.38.0", features = ["full"] }

[package.metadata.docs.rs]
all-features = true
rustdoc-args = ["--cfg", "docsrs"]

```

Set the `RUSTFLAGS` environment variable...
`set RUSTFLAGS=--cfg docsrs`

Then try to build the documentation...
`cargo +nightly doc --all-features`

The error list is long...
```
    Checking tokio v1.38.0
error[E0433]: failed to resolve: could not find `unix` in `os`
  --> C:\Users\bcook\.cargo\registry\src\index.crates.io-6f17d22bba15001f\tokio-1.38.0\src\io\bsd\poll_aio.rs:12:14
   |
12 | use std::os::unix::io::AsRawFd;
   |              ^^^^ could not find `unix` in `os`
   |
note: found an item that was configured out
  --> C:\Users\bcook\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib/rustlib/src/rust\library\std\src\os\mod.rs:26:9
   |
26 | pub mod unix {}
   |         ^^^^
note: found an item that was configured out
  --> C:\Users\bcook\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib/rustlib/src/rust\library\std\src\os\mod.rs:64:9
   |
64 | pub mod unix;
   |         ^^^^

error[E0433]: failed to resolve: could not find `unix` in `os`
  --> C:\Users\bcook\.cargo\registry\src\index.crates.io-6f17d22bba15001f\tokio-1.38.0\src\io\bsd\poll_aio.rs:13:14
   |
13 | use std::os::unix::prelude::RawFd;
   |              ^^^^ could not find `unix` in `os`
   |
note: found an item that was configured out
  --> C:\Users\bcook\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib/rustlib/src/rust\library\std\src\os\mod.rs:26:9
   |
26 | pub mod unix {}
   |         ^^^^
note: found an item that was configured out
  --> C:\Users\bcook\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib/rustlib/src/rust\library\std\src\os\mod.rs:64:9
   |
64 | pub mod unix;
   |         ^^^^

...
```

### PR DESCRIPTION
**Pull request by jofas**
I believe this fixes #6625 by only building the `io::bsd` module on Unix machines, in accordance with https://github.com/tokio-rs/tokio/issues/6625#issuecomment-2157740211. Putting the `#[cfg(unix)]` outside of the `cfg_aio!` macro felt less intrusive, as the macro is also used in `io::Interest`.[^1] I haven't yet tested this on Windows (or via cross-compilation).

While creating this PR I checked out the build logs from docs.rs and saw that the two [Windows builds](https://docs.rs/crate/tokio/1.41.0/builds/1400277/x86_64-pc-windows-msvc.txt) there are broken as well. I want to replicate how docs.rs builds the docs for Windows and fix them, too, so starting this PR as a draft. 

I wonder if building the docs on Windows would be a good idea to add to CI, to guard against future regressions?

[^1]: We should be able to build `Interest` on Windows with `--cfg docsrs` without problems.

### CODE REVIEW

**Comment by mox692** [comment]
> I wonder if building the docs on Windows would be a good idea to add to CI, to guard against future regressions?

Yes, this sounds good to me.

File: tokio/src/io/mod.rs [hunk 1]
```diff
@@ -231,6 +231,9 @@ cfg_io_driver_impl! {
     pub(crate) use poll_evented::PollEvented;
 }
 
+// The bsd module can't be build on Windows, so we completely ignore it, even
+// when building documentation.
+#[cfg(unix)]
```
Comment (jofas) [inline tokio/src/io/mod.rs:236]: I wonder if we should abstract this into a macro that better describes what we are doing? Something like `cfg_ignore_in_windows_docs!`?

**Comment by jofas** [comment]
With these changes, the windows docs can be  build with `-C docsrs` when cross-compiled in the [`crates-build-env`](https://github.com/rust-lang/crates-build-env) container. There is still a bunch of warnings about broken links in the docs though, like [this one](https://github.com/tokio-rs/tokio/blob/0dbdd196b6a9af9a106c654a8613088b81a64655/tokio/src/lib.rs#L258), for example. Should I fix those as well? If yes, what would be the best approach here? I probably would just configure out the parts of the docs that don't build on Windows. The main downside of this would be that I make the source of the docs less readable when I replace the nice `/// ...` lines with `cfg_attr!(unix, doc = ["..."])`, or such.

**Comment by Darksonn** [comment]
Let's not try to fix the broken links. Can we silence the warnings?

**Comment by jofas** [comment]
We could add `#![cfg_attr(windows, allow(rustdoc::broken_intra_doc_links)]` to silence the broken link warnings.

**Comment by jofas** [comment]
This should allow documentation to be buildable on Windows when `--cfg docsrs` is enabled, but I don't like it. The `#[cfg(docsrs)]`-related conditional compilation feels a bit fragmented to me as is and this PR as it currently stands makes matters worse IMO. Especially after I found the [`doc` module](https://docs.rs/tokio/latest/tokio/doc/index.html) from #3760, where someone already started mocking Windows-specific items that are part of the public interface. If we were to mock _all_ OS-specific types used throughout the library into such a module (not unlike the `loom` module), I think we could streamline conditional compilation and the quality of the documentation generated on Windows quite a bit. Would that be something the maintainers would be interested in?

On a side note, I'm not sure why the changes I made to the docs job of the CI workflow result in a `Expected — Waiting for status to be reported`  docs check, especially because the "all systems go" and "basic checks" jobs succeeded.

**Comment by mox692** [comment]
To me, current changes looks fine.

> On a side note, I'm not sure why the changes I made to the docs job of the CI workflow result in a `Expected — Waiting for status to be reported` docs check, especially because the "all systems go" and "basic checks" jobs succeeded.

Since this PR update the docs workflow, I guess the repository's [branch protection settings](https://github.com/orgs/community/discussions/26698) is relevant here?
@darksonn
Can you confirm this if you have time?

**Review by mox692 — APPROVED** [review]

**Comment by Darksonn** [comment]
I've updated the branch protection rules.

Regarding improving docs support on Windows ... I guess we could. But we would need to repeat all the hacks we use to make windows docs show up on Linux.

**Pull request merged**